### ISSUE TITLE
Support in browser support for editing files

### ISSUE THREAD

**Issue by paladox** [gitblit/gitblit#974]
Please support in browser editing of files and adding files please. Like gerrit has implanted this support gitblit should too. It would make everything easier.

**Comment by gitblit** [comment]
Hi paladox. Thanks for the idea. Somewhere in the imported issues there is an older request for the same thing. I won't be working on this feature any time soon. If you'd like to submit a PR with a proposed implementation I'll be happy to review.

**Comment by paulsputer** [comment]
:+1: This would be a useful feature to have.  It would be great to have something to quickly modify documents shown on the pages tab.  I had thought this may be considered differently for files and documents but I noticed a modern visual markdown editor [SimpleMDE](http://nextstepwebs.github.io/simplemde-markdown-editor/) uses codemirror behind the scenes so perhaps it's useful to consider both use-cases if selecting a client side library.

@gitblit would you be happy if I were to import [SimpleMDE](http://nextstepwebs.github.io/simplemde-markdown-editor/) into GitBlit?  If so I could do a PR for editing documents in the pages tab :)

**Comment by paladox** [comment]
@paulsputer yes please could you import at least it is some support for gitblit if you were to do that.

The editor could be made more advanced over time.

Thanks.

**Comment by gitblit** [comment]
@paulsputer Yup, that would be fine.  How about you work in my repo using pull requests rather than your fork.

**Comment by paulsputer** [comment]
Great stuff, many thanks @gitblit will do :)

**Comment by paladox** [comment]
Thanks.

**Comment by paulsputer** [comment]
I tried to add  (https://github.com/eirslett/frontend-maven-plugin) as a maven dependency via moxie to it would download npm seamlessly but couldn't get it to work.  @gitblit I don't have much experience with moxie config so you may have more success with getting the plugin working.

Currently node/npm must be installed manually on the development system (https://nodejs.org/en/).  This handles the JS dependencies CodeMirror and SimpleMDE and ensures only one version of CodeMirror is used. Perhaps this configuration with npm could also provide a starting point for future Gitblit 2.x developments...

**Comment by paladox** [comment]
@paulsputer thanks for working on this. Is it possible that a user does not have npm installed to use an alternative syntax highlighter code editor.

**Comment by paulsputer** [comment]
Hi @paladox npm is only required for building the gitblit source code and resources, it's not needed for viewing, so no worries there.  As long as users have a supported browser, as defined on the CodeMirror site, then they will have access to this functionality - that is pretty much any modern browser supporting javascript. ;)

There's still quite a bit of work required to get it to a production ready state, allowing changes to be committed back to the repository and handling merge conflicts gracefully are two aspects I'm moving onto, there maybe some more...

**Comment by paladox** [comment]
Ok thanks.

**Comment by paladox** [comment]
Does this support an option to disable the editor if gitblit acts as a mirror.

**Comment by paulsputer** [comment]
Referencing similar/duplicate request #635

**Comment by yurgis2** [comment]
Is there any active work on general file editing (not just documents)?

**Comment by paulsputer** [comment]
Currently not, I reckon there's a fair amount of work required to get it operating correctly and safely for general use outside of the document tab.  Until we can resolve merge conflicts in the browser I think it would be of limited value for code especially and we do begin to blur the line between repository/issue tracking and web a IDE.

I would anticipate the same `EditFile` page being used but adding in some switch to toggle between prosemirror and codemirror.  Doesn't look as if I'll have much time over the coming few months to work in this feature.  You're welcome to have a bash and submit a PR for review. :)
